Add binary sensors to Mazda integration

bdr99 opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
This PR adds binary sensor entities to the Mazda integration.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/21277

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [x] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

bdr99 committed:
Add binary sensors to Mazda integration

frenck commented on homeassistant/components/mazda/binary_sensor.py:
```diff
@@ -0,0 +1,184 @@
+    name_suffix: str
+
+    # Function to determine the value for this binary sensor, given the coordinator data
+    value: Callable[[dict], bool]
```
Can we make `dict` more specific?

bdr99 replied:
The `dict` here is the `DataUpdateCoordinator` data. See [here](https://github.com/bdr99/core/blob/3e6b5abe649bb5a5ec7bd6737b05a05fe6f4a76b/tests/components/mazda/fixtures/diagnostics.json#L8-L60) for an example of the coordinator data for this integration. This is a complex type which includes dicts from the pymazda external library. pymazda does not export typing info, so I'm not sure how to make `dict` more specific. Can you provide any advice/suggestions? Also, the current approach matches what is used in the sensor platform for this integration.

epenet replied:
At the very least I think you should make this a `dict[str, Any]` as you know the keys will be strings.

frenck commented on homeassistant/components/mazda/binary_sensor.py:
```diff
@@ -0,0 +1,184 @@
+    """Describes a Mazda binary sensor entity."""
+
+    # Function to determine whether the vehicle supports this binary sensor, given the coordinator data
+    is_supported: Callable[[dict], bool] = lambda data: True
+
+    # Function to return extra state attributes, given the coordinator data
+    state_attributes: Callable[[dict], dict] | None = None
```
Same as above, please make `dict` more specific

bdr99 replied:
Changed to `dict[str, Any]`.

frenck commented on homeassistant/components/mazda/binary_sensor.py:
```diff
@@ -0,0 +1,184 @@
+    return data["evStatus"]["chargeInfo"]["pluggedIn"]
+
+
+def _vehicle_is_charging(data):
+    """Determine if vehicle is charging."""
+    return data["evStatus"]["chargeInfo"]["charging"]
```
This isn't reused? Why not just put that directly into the lambda in the entity description?

(Applies to a bunch of more of these).

bdr99 replied:
I've converted functions to lambdas in cases where the lambda would be a single line. I was asked in https://github.com/home-assistant/core/pull/63423 to avoid multi-line lambdas, so I ignored the ones that would be multi-line lambdas.

bdr99 committed:
Convert functions to lambdas where lambdas would be a single line

bdr99 committed:
Improve typing for entity descriptions

epenet commented on homeassistant/components/mazda/manifest.json:
```diff
@@ -3,7 +3,7 @@
   "config_flow": true,
   "documentation": "https://www.home-assistant.io/integrations/mazda",
-  "requirements": ["pymazda==0.3.0"],
+  "requirements": ["pymazda==0.3.1"],
```
I know the change is very small, but if it is not directly required for the binary sensors I would argue it could (should?) be a separate PR.

bdr99 replied:
Sure, I'm fine with making it a separate PR. Reverted this change.

epenet commented on homeassistant/components/mazda/binary_sensor.py:
```diff
@@ -0,0 +1,165 @@
+    )
+
+
+def _doors_state_attributes(data):
+    """Get state attributes for the vehicle doors."""
+    door_status = data["status"]["doors"]
+
+    return {
+        "driver_door_open": door_status["driverDoorOpen"],
+        "passenger_door_open": door_status["passengerDoorOpen"],
+        "rear_left_door_open": door_status["rearLeftDoorOpen"],
+        "rear_right_door_open": door_status["rearRightDoorOpen"],
+    }
```
I can't quite decide if it is ok to have these as attributes? or if they should they be four separate sensors?

bdr99 replied:
Hmm, why wouldn't it be ok to have them as attributes? According to the developer docs, extra state attributes should be "information that further explains the state", which is what I think I've done here. The entity state says that the car has a door open, and the state attributes further explain that by specifying which door is open.

epenet replied:
I'll let @frenck or another review confirm, but we are generally moving away from attributes. If you look at the commit history you will find quite a few commits to "migrate attributes to individual sensors".
It's just that in this particular case I can't quite decide...

bdr99 replied:
Fair enough. The main reason I made it one sensor is because I thought it would make it easier for users to create automations. If they were four separate sensors then users would have to create an "or" condition every time. I was also trying not to clutter the UI with too many entities. But if another reviewer confirms that attributes should not be used here, I can change it to individual sensors.

DurgNomis-drol replied:
Giving my two cents, as i have made a custom integration for Toyota. I use individual sensors for each door. I can see where @bdr99 comes from, and I also thought long about what was the best option, as the user probably does not care _which_ door is open, but rather _if_ a door is open. My integration creates a lot of entities for just a single car, 25+. On the other hand, I personally prefer individual sensors over attributes, as attributes are harder to work with in the UI and in automations. (Though it have become easier over the last year).

bdr99 committed:
Change doors binary sensor to individual binary sensors for each door

bdr99 committed:
Update test fixtures

bdr99 commented:
I went ahead and changed the doors to individual binary sensors, since that is what other car integrations seem to be doing.

With this, I think I've addressed all review comments. @frenck and/or @epenet, mind taking another look at this PR?

bdr99 committed:
Remove charging binary sensor

bdr99 commented:
I removed the charging binary sensor from this PR because, as pointed out by @bdraco in https://github.com/home-assistant/core/pull/67597#discussion_r822128986, it makes more sense as a switch entity since it can be turned on and off.

Will open a separate PR to add the switch entity.

bdraco commented on homeassistant/components/mazda/binary_sensor.py:
```diff
@@ -0,0 +1,139 @@
+
+    entities: list[BinarySensorEntity] = []
+
+    for index, data in enumerate(coordinator.data):
+        for description in BINARY_SENSOR_ENTITIES:
+            if description.is_supported(data):
+                entities.append(
+                    MazdaBinarySensorEntity(client, coordinator, index, description)
+                )
+
+    async_add_entities(entities)
```
This could be a comprehension

bdr99 replied:
Fixed.

bdraco reviewed (changes requested):
Please resolve conflicts.  Thanks 👍

bdr99 committed:
Address review comment

bdr99 commented:
@bdraco Conflicts resolved.

bdraco commented on homeassistant/components/mazda/binary_sensor.py:
```diff
@@ -0,0 +1,135 @@
+    name_suffix: str
+
+    # Function to determine the value for this binary sensor, given the coordinator data
+    value: Callable[[dict[str, Any]], bool]
```
```suggestion
    value_fn: Callable[[dict[str, Any]], bool]
```

bdraco commented on homeassistant/components/mazda/binary_sensor.py:
```diff
@@ -0,0 +1,135 @@
+        name_suffix="Driver Door",
+        icon="mdi:car-door",
+        device_class=BinarySensorDeviceClass.DOOR,
+        value=lambda data: data["status"]["doors"]["driverDoorOpen"],
```
```suggestion
        value_fn=lambda data: data["status"]["doors"]["driverDoorOpen"],
```

bdraco commented on homeassistant/components/mazda/binary_sensor.py:
```diff
@@ -0,0 +1,135 @@
+        name_suffix="Passenger Door",
+        icon="mdi:car-door",
+        device_class=BinarySensorDeviceClass.DOOR,
+        value=lambda data: data["status"]["doors"]["passengerDoorOpen"],
```
```suggestion
        value_fn=lambda data: data["status"]["doors"]["passengerDoorOpen"],
```

bdraco commented on homeassistant/components/mazda/binary_sensor.py:
```diff
@@ -0,0 +1,135 @@
+        name_suffix="Rear Left Door",
+        icon="mdi:car-door",
+        device_class=BinarySensorDeviceClass.DOOR,
+        value=lambda data: data["status"]["doors"]["rearLeftDoorOpen"],
```
```suggestion
        value_fn=lambda data: data["status"]["doors"]["rearLeftDoorOpen"],
```

bdraco commented on homeassistant/components/mazda/binary_sensor.py:
```diff
@@ -0,0 +1,135 @@
+        name_suffix="Rear Right Door",
+        icon="mdi:car-door",
+        device_class=BinarySensorDeviceClass.DOOR,
+        value=lambda data: data["status"]["doors"]["rearRightDoorOpen"],
```
```suggestion
        value_fn=lambda data: data["status"]["doors"]["rearRightDoorOpen"],
```

bdraco commented on homeassistant/components/mazda/binary_sensor.py:
```diff
@@ -0,0 +1,135 @@
+        name_suffix="Trunk",
+        icon="mdi:car-back",
+        device_class=BinarySensorDeviceClass.DOOR,
+        value=lambda data: data["status"]["doors"]["trunkOpen"],
```
```suggestion
        value_fn=lambda data: data["status"]["doors"]["trunkOpen"],
```

bdraco commented on homeassistant/components/mazda/binary_sensor.py:
```diff
@@ -0,0 +1,135 @@
+        name_suffix="Hood",
+        icon="mdi:car",
+        device_class=BinarySensorDeviceClass.DOOR,
+        value=lambda data: data["status"]["doors"]["hoodOpen"],
```
```suggestion
        value_fn=lambda data: data["status"]["doors"]["hoodOpen"],
```

bdraco commented on homeassistant/components/mazda/binary_sensor.py:
```diff
@@ -0,0 +1,135 @@
+        name_suffix="Plugged In",
+        device_class=BinarySensorDeviceClass.PLUG,
+        is_supported=_plugged_in_supported,
+        value=lambda data: data["evStatus"]["chargeInfo"]["pluggedIn"],
```
```suggestion
        value_fn=lambda data: data["evStatus"]["chargeInfo"]["pluggedIn"],
```

bdraco commented on homeassistant/components/mazda/binary_sensor.py:
```diff
@@ -0,0 +1,135 @@
+    @property
+    def is_on(self):
+        """Return the state of the binary sensor."""
+        return self.entity_description.value(self.data)
```
```suggestion
        return self.entity_description.value_fn(self.data)
```

bdr99 committed:
Address review comments

bdr99 committed:
Update test_button.py

bdr99 commented:
@bdraco Resolved your latest round of comments.

bdraco commented:
I'll wait a few days to before merging to give other reviewers a chance to comment in case someone else sees something I missed.
